### ISSUE TITLE
Speed issues with Fractal Dimension

### ISSUE THREAD

**Issue by PedroACFerreira** [bonej-org/BoneJ2#196]
Hey there,

Currently very much enjoying working with BoneJ2! 
However, when running Fractal Dimension, it takes much more time than what I would expect for it to run when applying a small number of transitions. 
I'm trying to run FD on a 3D image (3D project of image stack with interpolation) - parameters "startBoxSize", 128, "smallestBoxSize", 1, "scaleFactor", 2, "translations" 3.

My setup has an Intel i7-9750h, 6 cores, 12 threads, and 32GB of RAM. Fiji uses about 30% of CPU and 35% of RAM while running Fractal Dimension with 3 transitions, and it still takes more than 15min for FD to finish. As I'm trying to run a batch script on 200+ files, this is a bit too much time.
Is this expected behavior? Is there something to be done, besides not doing any translations? 

Thank you very much!

**Comment by rimadoma** [comment]
Hi Pedro,

Thanks for submitting the issue. Adding translations _cubes_ the work needed, so with 3 you already have to do 27 times more calculations. This is because the sampling grid is moved 3 times in x, y and z dimensions. However, it's clear from the stats you submitted that my code is not efficient. I'll leave the issue open so that we can work on some speed ups. 

In the meanwhile, you could see if [Fraclac](https://imagej.nih.gov/ij/plugins/fraclac/FLHelp/Introduction.htm) or _Analyze > Tools > Fractal Dimension_ offer what you need. I'm not familiar with these tools, but they use different methods, and will give different results than Fractal Dimension in BoneJ.

**Comment by PedroACFerreira** [comment]
Thanks for the fast answer! Fraclac unfortunately only does box counting and not cube counting for 3D images from what I can gather, but I ended up just running BoneJ Fractal dimension with less translations, leaving the PC running a couple of nights should do the trick.
If there are any stats I can provide that might be helpful, let me know!

**Comment by rimadoma** [comment]
@PedroACFerreira Hi again, I managed to get Fractal Dimension with multiple translations to run faster. In my testing with a _100 x 100 x 100_ image filled with random pixels, the execution time dropped from _30 s_ to _8 s_. The speed up won't have any effect if you're using only 1 translation. Next time you see your ImageJ update `image-ops` , you'll get the faster version.

P.S. If you're willing to compromise on the _smallestBoxSize_ parameter, and use e.g. '2' or '4' pixels, you might find the calculations complete quicker

**Comment by rimadoma** [comment]
@mdoube, like mentioned in issue 616 on `imagej-ops`, I first tried running the individual boxes in the counting in parallel, but that offered only about 5 % speed up. I reckon it's because as the boxes get smaller, the overhead eats up any speed ups. Also, box counting is a  memory bound problem, because we're performing very simple operations on large number of pixels. That is, returns start diminishing fast after adding a few cores, because the bottle neck is in accessing the data.

It could be possible to try to run the op with chunks, where depending on the box size first a chunk might correspond to a box, and later a chunk might handle several boxes. It would be complicated to keep track of them though, especially since I generalised `BoxCount` to work in _n_-dimensions.

**Comment by rimadoma** [comment]
I realised yesterday that in some cases box counting performs unnecessary translations. For example, if box size is 2 * 2 * 2 you can only nudge it one pixel in any direction. Doing more than 1 'translation' is unnecessary.

I modified the box counting op so that it avoids unnecessary translations as box size gets smaller. That should further speed it up a bit.

**Comment by mdoube** [comment]
In general it's a good idea to avoid unnecessary work. I'd expect to get speedups up to about 8 - 10 cores/CPU threads, but you are right to think about how to minimise pixel array accesses. One way to do it is to add only the new pixels to your box (and remove the old ones) as it slides laterally. So if you have a 40 × 40 × 40 box moving one pixel in _x_, you pick up one face (1600 pixels) and reuse the rest (40³ - 1600 = 62400). Your box array should usually be small enough to fit in CPU cache. I did a small version of this in `[ConnectedComponents](https://github.com/bonej-org/BoneJ2/blob/0acc62a6f632574f5ffe739400d6ef833722e2e8/Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java#L718)`. The advantage should scale exponentially as box size gets bigger, but you do have to take care of potential out of bounds conditions.

**Comment by rimadoma** [comment]
> One way to do it is to add only the new pixels to your box (and remove the old ones) as it slides laterally

NB the implementation has never been a [sliding box count](https://en.wikipedia.org/wiki/Box_counting#Sliding_box_scans). What I enabled was the option to nudge the fixed grid _t_ times in each direction by _boxSize / (t + 1)_ pixels, where _t = number of translations_. That was already in the code of BoneJ1, just never used, because _t_ was always _0_.

> So if you have a 40 × 40 × 40 box moving one pixel in x, you pick up one face (1600 pixels)

I don't follow. The boxes never copy image data, they're [_Views_](https://javadoc.scijava.org/ImgLib2/net/imglib2/view/Views.html) to the same underlying image.

**Comment by mdoube** [comment]
> NB the implementation has never been a [sliding box count](https://en.wikipedia.org/wiki/Box_counting#Sliding_box_scans)

In which case ignore, because there are no redundant pixels from one box to the next.

> The boxes never copy arrays or image data, they're [_Views_](https://javadoc.scijava.org/ImgLib2/net/imglib2/view/Views.html) to the same underlying image.

There has to be _some_ degree of data copying, to get it off RAM and into the CPU for processing, regardless of the data structure used to access it. This strategy of sliding the neighbourhood (or View or whatever) is just a way to reduce slow RAM access when you already have the relevant information close to the CPU.

**Comment by mdoube** [comment]
> overhead eats up any speed ups. Also, box counting is a memory bound problem, because we're performing very simple operations on large number of pixels

Should be able to get up to 8-10× speedup. on typical workstation (and up to `nThreads`, usually 4, on a laptop). That much overhead worries me. Is there a lot of object creation and destruction? Are Views inefficient? One simple way of doing it is with the legacy `Multithreader` pattern, which uses a fixed pool of threads, all looking at a common `AtomicInteger` to get the next job from the queue. As long as you can calculate the box location from an integer, and because there is no writing in this algorithm, concurrency is minimal. This should max out RAM-CPU bandwidth and not leave it at 30%.

**Comment by rimadoma** [comment]
> Is there a lot of object creation and destruction?

Yes, when boxes are small and there's a lot of them. Can get rid of some of that, but I'll still need an array for each to keep track of location (see below).

> Are Views inefficient?

I don't know. In this case probably.

>  As long as you can calculate the box location from an integer
So `location = z * (w * h) + y * w + x`? The implementation in `imagej-ops` is n-dimensional. I'm not sure how to get around having an array for location, because you don't know how many dimensions you have. It has recursive for loops that pass the location array, until you get to the "bottom" and then access the pixel array.

**Comment by mdoube** [comment]
> > 
> 
> Yes, when boxes are small and there's a lot of them.

Can the objects be reused, by assigning new data to them?

> I'm not sure how to get around having an array for location
Do you mean that you pre-calculate the locations for all the boxes? That would work really well with an `AtomicInteger`: each thread just picks up the next location out of the array that is indicated by the ai counter

**Comment by rimadoma** [comment]
> Can the objects be reused, by assigning new data to them?

Object reuse / sharing wouldn't work with the way the implementation runs in parallel at the moment.

> Do you mean that you pre-calculate the locations for all the boxes?
I feel like we're talking in a different language again. Are you talking about running the pixel accesses _within_ a single box in parallel?

**Comment by mdoube** [comment]
> Object reuse / sharing wouldn't work with the way the implementation runs in parallel at the moment.
I usually give each thread its own copy of objects that can be reused, e.g. arrays that are used for neighbourhood lookups.

> Are you talking about running the pixel accesses _within_ a single box in parallel?
No. Each thread would get its own `View` or neighbourhood array or whatever structure you use to access the image data. Then the `View` / neighbourhood array would get moved (rather than destroyed & created), and the thread does its work on the data inside the `View` neighbourhood array, then at the end all the threads' work is added together.

**Comment by rimadoma** [comment]
@mdoube Had another look at FractalDimension. Testing with:
```
final Img<BitType> img = ArrayImgs.bits(256, 256, 256);
final double start = System.currentTimeMillis();
final List<ValuePair<DoubleType, DoubleType>> points = ops.topology().boxCount(img, 256L, 1L, 2.0);
final double end = System.currentTimeMillis();
final double time = (end - start) / 1000.0;
System.out.println("Time elapsed " + time + " s" + " results " + points.size());		
```
Execution time dropped from _25 s_ to _2.5 s_. PR here https://github.com/imagej/imagej-ops/pull/618. Using `Views.offsetInterval()` was the problem. 

Everything is still sequential. What runs in parallel is the multiple box grids, if you're using more than 0 translation. I'm still going to try running the individual boxes in parallel with something like:
```
final Callable<Long> task = () -> {
    long foreground = 0;
    final long[] boxStart = new long[interval.numDimensions()];
    final long[] position = new long[interval.numDimensions()];
    final RandomAccess<B> access = interval.randomAccess();
    while (boxer.hasNext()) {
        boxer.next(boxStart);
        if (hasForeground(access, boxStart, boxSize, position) {
            foreground++;
        }
    }
    return foreground;
}
final Future<Long> future = executor.submit(task);
...
private static final class Boxer {
    // Fills the array with the start position of the next box
    private synchronized void next(final long[] boxStart) { ... }
    private synchronized boolean hasNext { ... }
}
```
Like I mentioned above, I've tried this before, and it didn't offer speed ups. However, it was a different strategy. I don't think I can keep track of `AtomicInteger` and such while still holding on to the general N-dimensional implementation.

**Comment by rimadoma** [comment]
Another bit of good news is that I tested the above optimised N-dimensional version against a bare bones, 3D version. There was no speed difference.

**Comment by rimadoma** [comment]
I've parallised the BoxCount now. Execution time dropped further to _1.1 s_. 

Compared to the situation before PR 618 execution times have now dropped
* 256 x 256 x 256 empty image _25 s -> 1.1 s_
* 512 x 512 x 512 empty image _208 s -> 6 s_ 
* Parameters for both: `maxSize = 256`, `minSize = 1`, `scaling = 2.0 `

Findings:
* The above idea worked pretty much apart from the `synchronized` block creating a bottle neck. Instead of the threads sharing a `Boxer` they each have their own.
* I'm using the `Boxer` generators to get the next box coordinates as needed, because pre-calculating them all could take a lot of memory (512 x 512 x 512 x 3 x 8 bytes + array overhead...)
* Running single pixel boxes is more efficient with a single thread, so when boxSize gets < 2, `nThreads` is set _1_.

**Comment by rimadoma** [comment]
@PedroACFerreira 
Good news! When I first ran a _256 x 256 x 256_ empty image with your parameters ("startBoxSize", 128, "smallestBoxSize", 1, "scaleFactor", 2, "translations" 3), executing Fractal Dimension took _1789.195 s_ (almost 30 min). Now after all the speed ups I've worked in it takes _20.297 s_. You'll get the faster version the next time Fiji updates `imagej-ops.jar`.

If you don't want to wait that long (and are feeling adventurous), you can
1. Download [imagej-ops-0.45.3-SNAPSHOT.jar.zip](https://github.com/bonej-org/BoneJ2/files/4313042/imagej-ops-0.45.3-SNAPSHOT.jar.zip)
2. Close ImageJ if it's running
3. Unzip the JAR file and move it to your _Fiji.app/jar/_ folder
4. Save the original `imagej-ops.jar` file there somewhere else
5. Launch ImageJ and run Fractal Dimension
However, it's **critical** that you restore the original `imagej-ops.jar` as soon as possible. As long as the SNAPSHOT version is in your Fiji.app folder, it'll mess with your updates and potentially cause bugs.

**Comment by rimadoma** [comment]
BoneJ now uses `imagej-ops:0.45.3`, which has the faster versions of the _Fractal Dimension_ related algorithms.

**Comment by PedroACFerreira** [comment]
@rimadoma I've now been able to go and experiment with this again, so I came here just to let you know that this function of BoneJ is now indeed running incredibly faster! Roughly speaking, it took me about 15s-20 to run it (parameters "startBoxSize", 128, "smallestBoxSize", 1, "scaleFactor", 2, "translations", 5) on a 1024x1024x100 interpolated 3D image (a microscope image of a neuron in this case)! 
Would you like me to give you more detailed benchmarks on this? I should be running it again on 200+ files with a macro sometime in the next couple weeks as soon as I get my laptop thermals a bit lower, I could incorporate some logging in the macro and give you results!

**Comment by rimadoma** [comment]
@PedroACFerreira Glad to hear it's working faster! Bit of bad news though: one of the optimisations I coded has a bug, which may have affected your results. I'm sorry I cannot be certain either way - the issue has only appeared in specific environments, and I'm unable to reproduce it. Automated tests that pass on my computer and our Travis server fail on other machines. I have now submitted a fix for the issue. 

> Would you like me to give you more detailed benchmarks on this?
No need at the moment, thank you! Do let us know if you see a slow down after the above mentioned fix goes through (next version of imagej-ops). It's in a minor part of the whole Fractal Dimension process, so I don't expect it to have much effect.
